torch/nn/modules/module.py

```diff
@@ -1799,24 +1799,30 @@ def __delattr__(self, name):
             super().__delattr__(name)
 
     def _register_state_dict_hook(self, hook):
-        r"""Register a state-dict hook.
-
-        These hooks will be called with arguments: `self`, `state_dict`,
-        `prefix`, `local_metadata`, after the `state_dict` of `self` is set.
-        Note that only parameters and buffers of `self` or its children are
-        guaranteed to exist in `state_dict`. The hooks may modify `state_dict`
-        inplace or return a new one.
-        """
+        r"""See :meth:`~torch.nn.Module.register_state_dict_post_hook` for details."""
         handle = hooks.RemovableHandle(self._state_dict_hooks)
         self._state_dict_hooks[handle.id] = hook
         return handle
 
+    def register_state_dict_post_hook(self, hook):
+        r"""Register a post-hook for the :meth:`~torch.nn.Module.state_dict` method.
+
+        It should have the following signature::
+            hook(module, state_dict, prefix, local_metadata) -> None or state_dict
+
+        The registered hooks can modify the ``state_dict`` inplace or return a new one.
+        If a new ``state_dict`` is returned, it will only be respected if it is the root
```

Comment:
I guess removing this behavior is going to be tricky because it would require tracking which registration method was used to allow this or not?
Do you think it is worth doing to have full consistency? Or this is ok?

Reply:
Ok I updated this to track whether the hook was registered via the public or private method